kairos-contracts/demo-contract/contract/src/utils/events.rs

```diff
@@ -0,0 +1,11 @@
+use casper_event_standard::Event;
+extern crate alloc;
+use alloc::string::String;
+use casper_types::Key;
+
+#[derive(Event)]
+pub struct Deposit {
+    pub account: Key,
```

Comment:
The problem is that this could be an `AccountHash` or a `ContractPackageHash`.

If you follow how the code that creates the `Deposit` event you can see that `get_immediate_caller` is called to figure out who the depositor is. That function in turn is calling `call_stack_element_to_key` which returns an `Key` which is derived either from an `AccountHash` or `ContractPackageHash`.

Notice that the underlying representation of these two types is not identical.

The question is whether it makes sense to constrain it to `AccountHashes`, that would imply that `Contracts` could not have a kairos/L2 account.

We could probably wrap it in an `Either` type

Reply:
I did this on purpose to support both Account Hashs and Contract Hashs, the Key wrapper is my implementation and will work well for anything Kairos.

Reply:
@jonas089 If a contract is the depositor how do they make l2 transactions.

The `Key` is way to broad, it does not work.

To address an l2 account we cannot use casper account hash.
We need the user's public key.
It would be best if we got the public key from the l1 in the `Deposit` event.

If we cannot get the public key on the l1 in `Deposit` We could require the user to send it to the deposit endpoint with the deploy we forward.
But that would require going through the l2 to make a deposit which we might not want.

Making the l2 accounts Casper account hashes would require a lot of work and extra in zk compute.

```rust

#[repr(C)]
#[derive(PartialEq, Eq, Clone, Copy, PartialOrd, Ord, Hash)]
#[cfg_attr(feature = "datasize", derive(DataSize))]
pub enum Key {
    /// A `Key` under which a user account is stored.
    Account(AccountHash),
    /// A `Key` under which a smart contract is stored and which is the pseudo-hash of the
    /// contract.
    Hash(HashAddr),
    /// A `Key` which is a [`URef`], under which most types of data can be stored.
    URef(URef),
    /// A `Key` under which we store a transfer.
    Transfer(TransferAddr),
    /// A `Key` under which we store a deploy info.
    DeployInfo(DeployHash),
    /// A `Key` under which we store an era info.
    EraInfo(EraId),
    /// A `Key` under which we store a purse balance.
    Balance(URefAddr),
    /// A `Key` under which we store bid information
    Bid(AccountHash),
    /// A `Key` under which we store withdraw information.
    Withdraw(AccountHash),
    /// A `Key` variant whose value is derived by hashing [`URef`]s address and arbitrary data.
    Dictionary(DictionaryAddr),
    /// A `Key` variant under which system contract hashes are stored.
    SystemContractRegistry,
    /// A `Key` under which we store current era info.
    EraSummary,
    /// A `Key` under which we store unbond information.
    Unbond(AccountHash),
    /// A `Key` variant under which chainspec and other hashes are stored.
    ChainspecRegistry,
    /// A `Key` variant under which we store a registry of checksums.
    ChecksumRegistry,
}
```

Reply:
There is no way to derive the Public Key from an account hash.

I think your statement that "this doesn't work" is extreme. It definitely works and it might actually be the only solution, it just implies that we will have to hash Public Keys in some places and construct Account Hashs from those Public Keys when parsing the Events. 

This is not a problem and certainly does work. We can for simplicity restrict it to an account hash, but through callstack we will likely not be able tobretrieve a Public Key.

Reply:
The type `Key` does not work.
We could extract account hash and change the spec to use casper account hashes in the l2, but that is not ideal.

Could we take public key as a named arg and verify that the public key matches the account hash.
We could even leave the ladder off to support depositing into any l2 account.

Reply:
We could but then we would do the hashing inside the contract instead on the L2 which I think makes no sense.

Reply:
```rust
pub enum CallStackElement { Session { account_hash: AccountHash, }, StoredSession { account_hash: AccountHash, contract_package_hash: ContractPackageHash, contract_hash: ContractHash, }, StoredContract { contract_package_hash: ContractPackageHash, contract_hash: ContractHash, }, }
```

this is what we are working with, no way to derive the Public Key as far as I can tell.

Reply:
Get it via a named arg in deposit session.
We don't need to check it matches the caller's `AccountHash`.
Not checking it allows deposits into arbitrary accounts.
If we don't want that we can add a check latter.

Reply:
Haven't thought of it this way, good idea 👍

Reply:
I will rebase my changes on top of this PR, hopefully we could merge both info `feature/deposit-contract` branch.

Reply:
But this again is a Casper public key.

Reply:
We should make this decision asap @koxu1996 @Avi-D-coder so that we don't get blocked here. I'd like to merge this PR into the other contract PR because I think it's getting harder to work with when there are 2 PRs for the same issue.

Also we need the kairos-l1-utils crate which is not included on this branch for some reason. I am also confused over the issue that Andrzej pointed out where it looks like this branch is set to merge into Main.

<img width="255" alt="Screenshot 2024-05-10 at 22 50 34" src="https://github.com/cspr-rad/kairos/assets/49498646/93607273-2261-4284-9a32-d9d1b06e6445">

Reply:
I agree with Avi I think the easiest would be using the named arg.

@jonas089 @koxu1996  Setting this PR to merge into `master` serves the purpose of seeing what exact changes would make it in master obviously and to review how the bigger picture would look like.

It moreover triggers the whole pipeline to run since the GHA only run on PRs against master.

Reply:
@Avi-D-coder Okay, I will stick to non-tx deposit.

@marijanp PR is "deposit contract suggestions", so as the name suggests, it should include suggestions/improvements over contract, not the whole contract.